pubsub/google/cloud/pubsub_v1/publisher/batch/thread.py

```diff
@@ -0,0 +1,385 @@
+    batch.
+
+    If you are using this library, you most likely do not need to instantiate
+    batch objects directly; they will be created for you. If you want to
+    change the actual batching settings, see the ``batching`` argument on
+    :class:`~.pubsub_v1.PublisherClient`.
+
+    Any properties or methods on this class which are not defined in
+    :class:`~.pubsub_v1.publisher.batch.BaseBatch` should be considered
+    implementation details.
+
+    Args:
+        client (~.pubsub_v1.PublisherClient): The publisher client used to
+            create this batch.
+        topic (str): The topic. The format for this is
+            ``projects/{project}/topics/{topic}``.
+        settings (~.pubsub_v1.types.BatchSettings): The settings for batch
+            publishing. These should be considered immutable once the batch
+            has been opened.
+        autocommit (bool): Whether to autocommit the batch when the time
+            has elapsed. Defaults to True unless ``settings.max_latency`` is
+            inf.
+    """
+    def __init__(self, client, topic, settings, autocommit=True):
+        self._client = client
+
+        # These objects are all communicated between threads; ensure that
+        # any writes to them are atomic.
+        self._futures = []
+        self._messages = []
+        self._size = 0
+        self._settings = settings
+        self._status = self.Status.ACCEPTING_MESSAGES
+        self._topic = topic
+        self.message_ids = {}
+
+        # This is purely internal tracking.
+        self._thread = None
+
+        # Continually monitor the thread until it is time to commit the
+        # batch, or the batch is explicitly committed.
+        if autocommit and self._settings.max_latency < float('inf'):
+            self._thread = threading.Thread(target=self.monitor)
+            self._thread.start()
+
+    @property
+    def client(self):
+        """Return the client used to create this batch.
+
+        Returns:
+            ~.pubsub_v1.client.PublisherClient: A publisher client.
+        """
+        return self._client
+
+    @property
+    def messages(self):
+        """Return the messages currently in the batch.
+
+        Returns:
+            Sequence: The messages currently in the batch.
+        """
+        return self._messages
+
+    @property
+    def settings(self):
+        """Return the batch settings.
+
+        Returns:
+            ~.pubsub_v1.types.BatchSettings: The batch settings. These are
+                considered immutable once the batch has been opened.
+        """
+        return self._settings
+
+    @property
+    def size(self):
+        """Return the total size of all of the messages currently in the batch.
+
+        Returns:
+            int: The total size of all of the messages currently
+                 in the batch, in bytes.
+        """
+        return self._size
+
+    @property
+    def status(self):
```

Comment:
The multiprocessing version (now nuked from orbit) pulled it off a `manager.Namespace`. I could see this being distinctly different for different strategies.